src/Addons/CategoryFetcher.php

```diff
@@ -0,0 +1,240 @@
+ */
+
+namespace PrestaShop\Module\ProductComment\Addons;
+
+use PrestaShop\CircuitBreaker\AdvancedCircuitBreakerFactory;
+use PrestaShop\CircuitBreaker\Contracts\Factory;
+use PrestaShop\CircuitBreaker\Storages\SymfonyCache;
+use Symfony\Component\Cache\Simple\FilesystemCache;
+use Symfony\Component\CssSelector\CssSelectorConverter;
+use DOMDocument;
+use DOMXPath;
+use DOMNode;
+
+/**
+ * Class CategoryFetcher helps you to fetch an Addon category data. It calls the Addons
+ * API for name and link, and scrap the Addons platform to get its description.
+ * Every call is protected by a CircuitBreaker to avoid blocking the back-office.
+ */
+class CategoryFetcher
+{
+    const ADDONS_BASE_URL = 'https://addons.prestashop.com';
+    const ADDONS_API_URL = 'https://api-addons.prestashop.com';
+
+    const CLOSED_ALLOWED_FAILURES = 2;
+    const API_TIMEOUT_SECONDS = 0.6;
+
+    /**
+     * The timeout is longer for Addons platform as the content is bigger (HTML content)
+     */
+    const PLATFORM_TIMEOUT_SECONDS = 2;
+
+    const OPEN_ALLOWED_FAILURES = 1;
+    const OPEN_TIMEOUT_SECONDS = 1.2;
+
+    const OPEN_THRESHOLD_SECONDS = 60;
+
+    /** @var int */
+    private $categoryId;
+
+    /** @var array */
+    private $defaultData;
+
+    /** @var Factory */
+    private $factory;
+
+    /**
+     * @param int $categoryId
+     * @param array $defaultData
+     */
+    public function __construct(
+        $categoryId,
+        array $defaultData
+    ) {
+        $this->categoryId = $categoryId;
+        $this->defaultData = array_merge([
+            'id_category' => (int) $categoryId,
+        ], $defaultData);
+
+        $storage = new SymfonyCache(new FilesystemCache(
+            'addons_category',
+            0,
+            _PS_CACHE_DIR_
+        ));
+        $this->factory = new AdvancedCircuitBreakerFactory([
+            'closed' => [self::CLOSED_ALLOWED_FAILURES, self::API_TIMEOUT_SECONDS, 0],
+            'open' => [0, 0, self::OPEN_THRESHOLD_SECONDS],
+            'half_open' => [self::OPEN_ALLOWED_FAILURES, self::OPEN_TIMEOUT_SECONDS, 0],
+            'storage' => $storage,
+        ]);
+    }
+
+    /**
+     * @param string $isoCode Two letters iso code to identify the country (ex: en, fr, es, ...)
+     *
+     * @return array
+     */
+    public function getData($isoCode)
+    {
+        $category = $this->getCategoryFromApi($isoCode);
+        $category = $this->addTracking($category, $isoCode);
+        $category['description'] = $this->getDescription($category);
+
+        return $category;
+    }
+
+    /**
+     * @param string $isoCode
+     *
+     * @return array
+     */
+    private function getCategoryFromApi($isoCode)
+    {
+        $circuitBreaker = $this->factory->create([
+            'client' => [
+                'method' => 'POST',
+            ],
+        ]);
+        $apiJsonResponse = $circuitBreaker->call(
+            self::ADDONS_API_URL,
+            function () { return false; },
```

Comment:
you should return the same content type (json in this case) as documented :)

Reply:
I see what you mean but I'm not sure to understand
Why return an empty json, parse and then check that it's empty when the fallback return can be detected straight away by returning/checking the false value

Reply:
why not return a previous valid and cached response instead, if it exists ?

Reply:
the thing is the cache is managed by guzzlesubscriber, I need to see if it is possible to get back an expired cache